[0022]: Upgrade epoch SNARK data for flexibility, slashability and security (#67)
celo-org/celo-proposals: CIPs/0022.md

* `BHHash` - Bowe-Hopwood hash, as specified in the [Zcash protocol specification](https://github.com/zcash/zips/blob/master/protocol/protocol.pdf), on the Edwards curve defined over the same base field as BLS12-377, as specified in the [Zexe paper](https://eprint.iacr.org/2018/962.pdf).
* `Blake2Xs` - the Blake2s-based variant of Blake2X, defined in the [Blake2X paper](https://blake2.net/blake2x.pdf).
* `pad(array, n, padding_element)` - enlarge the array to size `n`, adding `padding_element` in the new places.
* `truncate(array, n)` - truncate the array to the first `n` elements.

The encoding is currently implemented as:

```
Blake2Xs(
    BHHash(
        encode_integer_16_le(epoch number) || 
        encode_integer_32_le(maximum_non_signers) || 
        validator set
    )
)
```

where `validator_set` is the concatenation of the results of `encode_public_key_le` on each validator public key.

We propose to change it to the following. We first define the parameters `MAX_VALIDATORS = 150` and `SEC_PARAM = 128`. We then specify the new encoding.

```
Blake2Xs(
    encode_integer_16_le(epoch number) || 
    BHHash(

>         bytes_le_to_bits_le(truncate(parent_hash), SEC_PARAM/8) ||

Reviewer: The closing parenthesis for truncate is misplaced, this line should instead say:
```
        bytes_le_to_bits_le(truncate(parent_hash, SEC_PARAM/8)) ||
```

Author: Oops, thanks!

        encode_integer_32_le(maximum_non_signers) || 
        pad(validator set, MAX_VALIDATORS, G1_GENERATOR)
    )
)
```

> where `validator_set` is the concatenation of the results of `encode_public_key_le` on each validator public key, `parent_hash` is the hash of the parent block header and

Reviewer: I think it'd be good to clarify that parent block here refers to the block that's one epoch prior (the parent in the chain of epoch blocks, not in the chain of blocks in general).  It's logical in the context of lightest sync and Plumo, where only epoch blocks are involved, but in general when I hear "parent block" I think of the previous block in the blockchain.

Author: Very good point. What do you think about the new phrasing?

Reviewer: Looks good to me.

The currently deployed version of the encoding can be improved:
1. It only works correctly when the validator set is at its maximum size. 
2. As it's designed to be SNARK-friendly, the encoding uses the Bowe-Hopwood SNARK-friendly hash function on the entire elected validator set. As a result, showing a double-signing proof currently requires hashing the entire elected validator set with a non EVM-friendly hash, having too high gas costs.
3. It's vulnerable to a "future subcommittee" attack, where an attacker that compromises enough validators can pre-sign encodings with future epoch numbers and use them when their compromised subcommittee is elected in the future.

This CIP upgrades the encoding to address all three topics.

## Motivation

- Support validator sets which are not full up to a specified maximum.

> - Create the necesasry infrastructure for on-chain slashing in the case of double-signing.

Reviewer 2: ```suggestion
- Create the necessary infrastructure for on-chain slashing in the case of double-signing. 
```

How about a bullet for the future subcommittee attack?

## Specification

We define a few helper functions:
* `encode_integer_X_le` - encodes an unsigned_integer having X bits to bits in a little-endian order.
* `encode_public_key_le` - encodes a BLS12-377 point to a compressed form in bits - the little-endian bits of the x coordinate and a single bit whether y is the greatest of the two possibilities.
* `bits_be_to_bytes_le` - converts big-endian bits to little-endian bytes.
* `bytes_le_to_bits_be` - converts little-endian bytes to big-endian bits.
* `bytes_le_to_bits_le` - converts little-endian bytes to little-endian bits.
* `BHHash` - Bowe-Hopwood hash, as specified in the [Zcash protocol specification](https://github.com/zcash/zips/blob/master/protocol/protocol.pdf), on the Edwards curve defined over the same base field as BLS12-377, as specified in the [Zexe paper](https://eprint.iacr.org/2018/962.pdf).
* `Blake2Xs` - the Blake2s-based variant of Blake2X, defined in the [Blake2X paper](https://blake2.net/blake2x.pdf).

> * `pad(array, n, padding_element)` - enlarge the array to size `n`, adding `padding_element` in the new places.
> * `truncate(array, n)` - truncate the array to the first `n` elements.

Reviewer 2: ```suggestion
* `pad(array, n, padding_element)` - enlarges the array to size `n`, adding `padding_element` in the new places.
* `truncate(array, n)` - truncates the array to the first `n` elements.
```

Adding the current and parent block header hashes prevents the "future subcommittee" attack since it adds an unpredictable value that the attacker can't predict in advance and therefore can't pre-sign, and establishes a chain of epoch headers. Specifically, the header contains the state hash, which in turn contains the result of the Celo randomness beacon. We take only `SEC_PARAM/8` bytes, since, assuming the bytes are sufficiently unpredictable, an attacker would need to pre-sign around `2^128` different messages in order to succeed.

## Test Cases

* **WIP** - We provide the [plumo-prover](https://github.com/celo-org/plumo-prover) codebase, which consumes data from a Celo deployment and generates Plumo proofs. It will be run on private testnets before the test period and on Alfajores and Baklava during the test period.
* **WIP** - We provide a proof-of-concept implementation of a slashing contract that utilizes [CIP20](https://github.com/celo-org/celo-proposals/compare/prestwich/cip-0020) for Blake2Xs and [CIPX]() for BLS12-377 operations.

## Implementation

> * **WIP** - The changes are implemented as PRs on [celo-blockhain](https://github.com/celo-org/celo-blockchain) ([#1158](https://github.com/celo-org/celo-blockchain/pull/1158)), [celo-bls-snark-rs](https://github.com/celo-org/celo-bls-snark-rs) ([#183](https://github.com/celo-org/celo-bls-snark-rs/pull/183)) and [celo-bls-go](https://github.com/celo-org/celo-bls-go).

Reviewer 2: ```suggestion
* **WIP** - The changes are implemented as PRs on [celo-blockchain](https://github.com/celo-org/celo-blockchain) ([#1158](https://github.com/celo-org/celo-blockchain/pull/1158)), [celo-bls-snark-rs](https://github.com/celo-org/celo-bls-snark-rs) ([#183](https://github.com/celo-org/celo-bls-snark-rs/pull/183)) and [celo-bls-go](https://github.com/celo-org/celo-bls-go).
```

## Deployment Considerations

* This CIP can in fact be deployed as a soft-fork, in the sense that only validators MUST upgrade. Full nodes and light-clients do not verify the correctness of this field. Nevertheless, we suggest to deploy it in a hard-fork to minimize the amount of upgrades validators have to perform.

> * The CIP puts hard upper-bound on the number of validators to be 150.

Reviewer 2: ```suggestion
* The CIP puts a hard upper-bound on the number of validators to be 150.
```

        )
    )
)
```

where `counter` is the first number where `encode_to_point` succeeds starting from `0` and at most `255`, `validator_set` is the concatenation of the results of `encode_public_key_le` on each validator public key, `epoch_hash` is the hash of the current epoch block header, `parent_hash` is the hash of the previous epoch block header (17280 blocks in the past) and

`G1_GENERATOR = (81937999373150964239938255573465948239988671502647976594219695644855304257327692006745978603320413799295628339695, 241266749859715473739788878240585681733927191168601896383759122102112907357779751001206799952863815012735208165030)`

## Rationale

A single Plumo proof can support a fixed amount of validators and epochs. We believe that 150 validators and 130 epochs (~= 130 days) will satisfy Celo's needs for the next 2 years.

Moving the epoch number outside the `BHHash` allows for an efficient on-chain double-signing proof - this proof doesn't need the information inside the `BHHash`, it's just required to show there are two aggregate signatures by a super majority on the same epoch number and distinct opaque outputs of `BHHash`.

> Adding the current and parent block header hashes prevents the "future subcommittee" attack since it adds an unpredictable value that the attacker can't predict in advance and therefore can't pre-sign, and establishes a chain of epoch headers. Specifically, the header contains the state hash, which in turn contains the result of the Celo randomness beacon. We take only `SEC_PARAM/8` bytes, since, assuming the bytes are sufficiently unpredictable, an attacker would need to pre-sign around `2^128` different messages in order to succeed.

Reviewer: Some more details about the future subcommittee attack would be good.  For one thing, I had a hard time finding information about it by googling it.  But, also, at first I thought since the entropy includes two header hashes, we only need to take `SEC_PARAM/16` bytes from each.  Then I thought maybe not, because the attack could perhaps be done during one epoch, so the parent hash could be know.  But then it wasn't clear to me why there should be `SEC_PARAM/8` bytes from the parent hash.  Is it to ease the SNARK by having the two hashes be the same length?  Regardless, I guess my point is just that more detail would be good on these points, both about the attack (can be a link to details elsewhere), and about the requirements for the chain of epoch headers that are behind the choice to include `SEC_PARAM/8` for the parent hash too.

> # CIP [0022]: Upgrade epoch SNARK data for flexibility, slashability and security

Reviewer 3: Hi @kobigurk I'd remove this title as the preamble will have all the information needed. Preamble will scraped later with a CIP website that I will integrate later.

---
cip: 22
title: Upgrade epoch SNARK data for flexibility, slashability and security
author: Kobi Gurkan <me@kobi.one>

> discussions-to: <URL>

Reviewer 3: Can you add a discussions-to url? It can be as easy as opening a Github ticket that is called: "Discussions to  Upgrade epoch SNARK data for flexibility, slashability and security" and linking the URL here. Or, if you prefer, you can make one on the forum.celo.org.
